src/Analyzers/CSharp/Analyzers/UseUTF8StringLiteral/UseUTF8StringLiteralDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,204 @@
+
+                context.RegisterOperationAction(c => AnalyzeOperation(c, expressionType), OperationKind.ArrayCreation);
+            });
+
+        private void AnalyzeOperation(OperationAnalysisContext context, INamedTypeSymbol? expressionType)
+        {
+            var arrayCreationOperation = (IArrayCreationOperation)context.Operation;
+
+            // Don't offer if the user doesn't want it
+            var option = context.GetOption(CSharpCodeStyleOptions.PreferUTF8StringLiterals);
+            if (!option.Value)
+                return;
+
+            // Only replace arrays with initializers
+            if (arrayCreationOperation.Initializer is null)
+                return;
+
+            // Using UTF8 string literals as nested array initializers is invalid
+            if (arrayCreationOperation.DimensionSizes.Length > 1)
+                return;
+
+            // Must be a byte array
+            if (arrayCreationOperation.Type is not IArrayTypeSymbol { ElementType.SpecialType: SpecialType.System_Byte })
+                return;
+
+            // UTF8 strings are not valid to use in attributes
+            if (arrayCreationOperation.Syntax.Ancestors().OfType<AttributeSyntax>().Any())
+                return;
+
+            // Can't use a UTF8 string inside an expression tree.
+            var semanticModel = context.Operation.SemanticModel;
+            Contract.ThrowIfNull(semanticModel);
+            if (arrayCreationOperation.Syntax.IsInExpressionTree(semanticModel, expressionType, context.CancellationToken))
+                return;
+
+            var elements = arrayCreationOperation.Initializer.ElementValues;
+
+            // If the compiler has constructed this array creation, then we don't want to do anything
+            // if there aren't any elements, as we could just end up inserting ""u8 somewhere.
+            if (arrayCreationOperation.IsImplicit && elements.Length == 0)
+                return;
+
+            if (!TryConvertToUTF8String(builder: null, elements))
+                return;
+
+            if (arrayCreationOperation.Syntax is ImplicitArrayCreationExpressionSyntax or ArrayCreationExpressionSyntax)
+            {
+                ReportArrayCreationDiagnostic(context, arrayCreationOperation.Syntax, option.Notification.Severity);
+            }
+            else if (elements.Length > 0 && elements[0].Syntax.Parent is ArgumentSyntax)
+            {
+                // For regular parameter arrays the code fix will need to search down
+                ReportParameterArrayDiagnostic(context, arrayCreationOperation.Syntax, elements, option.Notification.Severity, ArrayCreationOperationLocation.Descendants);
+            }
+            else if (elements.Length > 0 && elements[0].Syntax.Parent.IsKind(SyntaxKind.CollectionInitializerExpression))
+            {
+                // For collection initializers where the Add method takes a parameter array, the code fix
+                // will have to search up
```

Comment:
so in general i don't likw the idea that the fix will have to 'search'.  better to have the analyzer find the excact node to be fixed and put that into the diagnostic (in 'additionallocations').  If hte analzyer can't find the exact node, then nothing is reported.  that way we don't potentially end up with a mismatch where the analyzer hopes everything is fixable, but it turns out not to be when the fixer runs (And we crash or offer nothing).